server/service/src/pack_unit.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::HashMap;
+
+use repository::{
+    EqualFilter, PackUnitRow, PackUnitRowRepository, RepositoryError, StockLineFilter,
+    StockLineRepository, StockOnHandFilter, StockOnHandRepository, StorageConnection,
+};
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PackUnit {
+    pub item_id: String,
+    pub most_used_pack_unit_id: String,
+    pub pack_units: Vec<PackUnitRow>,
+}
+
+pub fn get_pack_units(
+    connection: &StorageConnection,
+    store_id: &str,
+) -> Result<Vec<PackUnit>, RepositoryError> {
+    let stock_lines = StockLineRepository::new(connection)
+        .query_by_filter(StockLineFilter::new(), Some(store_id.to_string()))?;
```

Comment:
This query is probably the most expensive one because it also returns joined rows which are not needed. So this could be optimized if needed, i.e. have a repo or repo method to only return the item_id + pack_size or a joined pack_unit + pack_size query...

Regarding your original question. You could create a view that gives you the data in a nicer format and do the rest of the logic here. The view columns could be like:

|item_id|pack_size|sum_available_stock_on_hand|total_number_of_stock_lines|

where the view is grouped by item_id and pack_size. This might be not too complicated. 

However, I am a bit with with Mark, it is probably simpler to understand when leaving it as it is, i.e. all logic in one place.